pkg/controller/virtualmachine/virtualmachine.go

```diff
@@ -0,0 +1,462 @@
+			count++
+		}
+	}
+
+	if count != 1 {
+		return nil, fmt.Errorf("expected to find 1 pod, but found %d associated with vmi %s", count, vmi.Name)
+	}
+
+	return &runningPod, nil
+}
+
+func (h *Handler) getPodEnv(pod *corev1.Pod) ([]byte, error) {
+	e, err := executor.NewRemoteCommandExecutor(h.ctx, h.config, pod)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up remote executor for pod %s: %v", pod.Name, err)
+	}
+	return e.Run("env", []string{})
+}
+
+// convertEnvToMap converts env info to a map to make it easier to find device
+// allocation details
+func convertEnvToMap(podEnv []byte) (map[string]string, error) {
+	var contents []string
+	bufReader := bufio.NewReader(bytes.NewReader(podEnv))
+
+	for {
+		line, err := bufReader.ReadBytes('\n')
+		if len(line) != 0 {
+			contents = append(contents, strings.TrimSuffix(string(line), "\n"))
+		}
+
+		if err != nil {
+			if err != io.EOF {
+				return nil, err
+			}
+			break
+		}
+	}
+
+	resultMap := make(map[string]string)
+	for _, line := range contents {
+		lineArr := strings.Split(line, "=")
+		resultMap[lineArr[0]] = strings.Join(lineArr[1:], "=")
+	}
+	return resultMap, nil
+}
+
+// check and clear deviceAllocation annotations if needed
+func (h *Handler) checkAndClearDeviceAllocation(vmi *kubevirtv1.VirtualMachineInstance) error {
+	vmObj, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return fmt.Errorf("error fetching vm %s from cache: %s", vmi.Name, err)
+	}
+
+	_, ok := vmObj.Annotations[v1beta1.DeviceAllocationKey]
+	// no key, nothing is needed
+	if !ok {
+		return nil
+	}
+	delete(vmObj.Annotations, v1beta1.DeviceAllocationKey)
+	_, err = h.vmClient.Update(vmObj)
+	return err
+}
+
+func buildVGPUMap(vgpuDevices []*v1beta1.VGPUDevice) map[string]string {
+	result := make(map[string]string)
+	for _, vgpu := range vgpuDevices {
+		if vgpu.Status.UUID != "" {
+			result[vgpu.Status.UUID] = vgpu.Name
+		}
+	}
+	return result
+}
+
+func buildPCIDeviceMap(pciDevices []*v1beta1.PCIDevice) map[string]string {
+	result := make(map[string]string)
+	for _, device := range pciDevices {
+		result[device.Status.Address] = device.Name
+	}
+	return result
+}
+
+func generateAllocationDetails(hostDeviceMap, gpuMap map[string][]string) *v1beta1.AllocationDetails {
+	resp := &v1beta1.AllocationDetails{}
+	if len(hostDeviceMap) > 0 {
+		hostDeviceMap = dedupDevices(hostDeviceMap)
+		resp.HostDevices = hostDeviceMap
+	}
+
+	if len(gpuMap) > 0 {
+		gpuMap = dedupDevices(gpuMap)
+		resp.GPUs = gpuMap
+	}
+	return resp
+}
+
+// if multiple devices of same type are added to a VM, there may be duplicates in the deviceDetails values
+// since env variable would have been looked up twice from envMap and added to deviceDetails
+// so we dedup the unique device names
+func dedupDevices(deviceMap map[string][]string) map[string][]string {
+	for key, val := range deviceMap {
+		deviceMap[key] = slices.Compact(val)
+	}
+	return deviceMap
+}
```

Comment:
Jack and I found the duplicate in the annotation, it seems not expected?
![image_720](https://github.com/user-attachments/assets/96424be9-f6ea-47c1-bb29-3c5867140f57)

Reply:
FYI, it happened when we attached two vGPU devices to a VM.

Reply:
i can fix it up

Reply:
i have identified the reason why the names were being duplicated. I have now fixed this up and the additional commit contains the fix. I have kept the commits separate different to make it easier to track the changes. I will squash them once we are happy with the new set of changes.